[WIP] Various IDE improvements and new features

cedihegi opened the pull request:
Changes to support various new features in prusti-assistant. 

@Aurel300 
@jthomme1

Commit:
collecting information

Commit:
some updates, had a lot of problems with just passing flags...

Commit:
passing method calls and verification items to ide

Commit:
fixed duplicate method calls
(this commit is not compiling however!)

Commit:
Fixed the issue of no_verify causing subsequent verifications to fail by simply creating a fake error in the case
that no_verify and show_ide_info are set.

Commit:
now finding correct defpath for function calls and also handling binops, even though they introduce some new problems

Commit:
Fixed a bug for calls where the DefId can not be found apparently.

Commit:
fixed minor bug and formatting

Commit:
now finding defpath's for MethodCalls too

Commit:
Figured out how to resolve MethodCalls, but we are still in trouble..

Commit:
Commit finally to own fork

Commit:
Finally catching all relevant method calls as far as I can tell..

Commit:
initial quantifier instantiations working

Commit:
generating signatures correctly in some cases, struggling with traitbounds

Commit:
successfully resolving projections (it seems)

Commit:
started on generating external specs for traits

Commit:
WIP: stream messages from server to client

Commit:
Managed to subst EarlyBinders

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into proto-extspec

Commit:
upgraded my parts to current rust version

Commit:
Asynchronously report messages from the server to the client via WebSocket. Report quantifier instantiations split by methods.

Commit:
refactored extern spec generation to try and make it more readable.

Commit:
renamed stuff

Commit:
Merge branch 'joseph_quant' into proto-extspec

Commit:
Merge remote-tracking branch 'upstream/master' into proto-extspec

Commit:
Adjusted viper VerificationResult so we can passe times, and whether is was cached to clients.

Commit:
WIP: restructure the process verification a bit

Commit:
Fixed warnings and a bug.

Commit:
renamed a file and adjusted text for IDE information output

Commit:
Finish process_verification restructuring. Add a parameter for qi.profile_freq. Encode the quantifier ID in the line of the position in order to not having to modify the silver way of encoding a quantifier.

Commit:
Merge remote-tracking branch 'cedric/proto-extspec' into quant

Commit:
Changed message for fake errors and adjusted names of CompilerInfo procedures.

Commit:
Merge remote-tracking branch 'jthomme1/quant' into proto-extspec

Commit:
Add QuantifierChosenTriggersMessage reporting

Commit:
Merge remote-tracking branch 'cedric/proto-extspec' into quant

Commit:
passing information about specifications of calls to IDE now, called EncodingInfo

Commit:
WIP: async processing of verification info

Commit:
(IDE) Contract Spans are now also collected for pedges and termatinations (altough ranges are not always accurate)

Commit:
Cleaning up unused imports and debugging print statements

Commit:
Some minor output changes

Commit:
Merge cedric's branch

Commit:
Minor fixes

Commit:
Merge remote-tracking branch 'upstream/master' into quant

Commit:
give spans of pledges to IDE and filter out trusted methods and predicates from items offered for selective verification

Commit:
Merge remote-tracking branch 'joseph/quant' into proto-extspec

Commit:
updated ide-parts to new compiler version

Commit:
Emit everything in a compiler note/message

Commit:
Make every token CamelCase

Commit:
Set config variable for viper message reporting to default false.

Commit:
Proper camelCase messages

Commit:
Can not automatically generate extern_specs for local methods anymore

fpoli commented on prusti-server/src/process_verification.rs:
```diff
@@ -4,24 +4,110 @@
+// one structure that lives for all the requests and has a single thread working on all the
+// requests sequentially
+// on reception of a verification request, we send it through a channel to the already running
+// thread
```
Use `///` for [doc comments](https://doc.rust-lang.org/rust-by-example/meta/doc.html)

fpoli commented on viper/src/verification_result.rs:
```diff
@@ -6,20 +6,42 @@
 /// The result of a verification request on a Viper program.
 #[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
-pub enum VerificationResult {
+pub enum VerificationResultType{
```
The `...Kind` suffix is a common choice in Rust to avoid naming `...Type` a type.

fpoli commented on prusti-viper/src/encoder/encoder.rs:
```diff
@@ -103,6 +104,7 @@ pub struct Encoder<'v, 'tcx: 'v> {
     /// this requires special care when encoding array/slice accesses which may come with
     /// bound checks included in the MIR.
     pub(super) is_encoding_trigger: Cell<bool>,
+    pub spans_of_call_contracts: RefCell<Vec<SpanOfCallContracts>>,
```
Can you add a doc comment?

fpoli commented on prusti-interface/src/prusti_error.rs:
```diff
@@ -40,6 +40,7 @@ pub enum PrustiErrorKind {
     Warning,
     /// A warning which is only shown if at least one error is emitted.
     WarningOnError,
+    Message,
```
If there are no Prusti functions that expect `PrustiErrorKind` to always be causes by an error, at this point I think we should rename `PrustiErrorKind` to something like `PrustiDiagnosticKind`.

fpoli commented on prusti-smt-solver/src/context.rs:
```diff
@@ -15,6 +16,7 @@ pub(super) struct Context {
 
 impl Context {
     pub(crate) async fn new() -> Self {
+        trace!("BBBB");
```
A leftover to remove, I guess?

fpoli commented on prusti-viper/src/ide/encoding_info.rs:
```diff
@@ -0,0 +1,46 @@
+use super::vsc_span::VscSpan;
+/// stores the spans of a calls contracts.
+/// obtained during encoding
+
```
Put an empty line before the doc comment and remove the empty line after.

Commit:
Fix GlobalRef detaching warning. Add a few comments. Remove some debugging leftovers.

fpoli commented on prusti/src/callbacks.rs:
```diff
@@ -127,8 +133,42 @@ impl prusti_rustc_interface::driver::Callbacks for PrustiCompilerCalls {
+                    };
+                    verify(env, def_spec, verification_task);
+                } else {
+                    let target_def_path = config::selective_verify().unwrap();
```
Consider if it's better to rewrite the branch to `if let Some(..) = config::selective_verify() { .. } else { .. }` to avoid the dependency between `is_none()` and `unwrap()`.

fpoli commented on prusti-server/src/server.rs:
```diff
@@ -41,61 +41,53 @@ pub fn spawn_server_thread() -> SocketAddr {
     receiver.recv().unwrap()
 }
 
+// initialize on first access
```
The comment is implied by the `Lazy` type. Can you instead summarize what this global does?

fpoli reviewed (commented):
See the comments.

Commit:
Better comment for the VerificationRequestProcessing global variable.

Commit:
Remove leftover comment.

Commit:
Merge remote-tracking branch 'jthomme1/quant'

Commit:
Correct formatting and removed all clippy warnings except for 2

Commit:
Fix last 2 clippy warnings

Commit:
Updated viper-toolchain, added various comments / documentation, addressed a few reviews

Commit:
corrected tag for recent viper-ide nightly release

Commit:
updated viper-toolchain again, this time to a compatible version

Commit:
fixed error due to new scala version

Commit:
made testcases compile at least, but one is hard to update to new structure of verifier

Commit:
Fixed prusti-server test. Why do we get a 255 exit code on a successful verification?

Commit:
Adjusted ui testcase that has additional note about existential quantifier now

Commit:
Bugfixes for 2 testcases. All tests should be passing now

Commit:
Add debug output

Commit:
Merge remote-tracking branch 'cedric/master' into quant

Commit:
Merge remote-tracking branch 'jthomme1/quant'

Commit:
Make cache save when not using a server

Commit:
Merge remote-tracking branch 'jthomme1/quant'

Commit:
Gracefully close also the receiving end of the websocket.

Commit:
Close the websocket connection correctly

Commit:
Merge remote-tracking branch 'jthomme1/quant'

Commit:
Fix bug where successful selective verification is still cached

In general, when skipping verification, we introduced a so called
fake_error to make sure the compiler does not just cache this
as successful verification and then skip later verifications even
if that's not intended. Same issue occured for selective verification,
when verifying a single method that is "correct", this would be cached.
Now we throw a fake error in this case too, to avoid this.

Commit:
Fix bug in translation of spans to vscode-spans

Commit:
Bump prusti version to 0.3.0

Prusti Assistant offers a few new features, and it checks for this version
for them to be available.

Commit:
Update viper-toolchain version

Aurel300 commented on viper/src/verification_result.rs:
```diff
@@ -6,9 +6,30 @@
+
+impl VerificationResult {
+    pub fn is_success(&self) -> bool {
+        matches!(self.result_type, VerificationResultKind::Success)
```
```suggestion
        self.result_type.is_success()
```

Aurel300 commented on viper/src/verification_result.rs:
```diff
@@ -6,9 +6,30 @@
+#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct VerificationResult {
+    pub item_name: String,
+    pub result_type: VerificationResultKind,
```
```suggestion
    pub kind: VerificationResultKind,
```

Aurel300 commented on prusti/src/verifier.rs:
```diff
@@ -18,11 +22,6 @@ pub fn verify(env: Environment<'_>, def_spec: typed::DefSpecificationMap) {
         debug!("Prepare verification task...");
         // TODO: can we replace `get_annotated_procedures` with information
         // that is already in `def_spec`?
```
Please move the TODO comment to where you now invoke `get_annotated_procedures_and_types`.

Aurel300 commented on prusti-utils/src/config.rs:
```diff
@@ -102,6 +102,9 @@ lazy_static::lazy_static! {
         settings.set_default("assert_timeout", 10_000).unwrap();
         settings.set_default("smt_qi_eager_threshold", 1000).unwrap();
+        settings.set_default("smt_qi_profile", true).unwrap();
+        settings.set_default("smt_qi_profile_freq", 10000).unwrap();
```
Will these be passed to Z3 when not using the IDE? And in such a case, will the messages not be reported to the user? If so, QI profiling should be disabled by default.

Aurel300 replied:
(After looking at how they are used in the other file:) please make these be `Option<_>`, such that they are only passed as arguments when a value is set. The IDE can set these if they are needed.

Aurel300 commented on prusti-utils/src/config.rs:
```diff
@@ -1017,3 +1043,29 @@ pub fn cargo_command() -> String {
+/// Used for selective verification, can be passed a String containing
+/// the DefPath of the method to be verified
+pub fn selective_verify() -> Option<String> {
+    read_setting("selective_verify")
```
Rename to `verify_only_defpath` to be consistent with `verify_only_basic_block_path`.

Aurel300 commented on prusti-utils/src/config.rs:
```diff
@@ -102,6 +102,9 @@ lazy_static::lazy_static! {
         settings.set_default("quiet", false).unwrap();
```
Make sure to update the `flags.md` file in the dev guide (`docs/dev-guide` in the repo) with all the flags you've added.

Aurel300 commented on prusti-utils/src/config.rs:
```diff
@@ -494,6 +504,22 @@ pub fn smt_qi_eager_threshold() -> u64 {
     read_setting("smt_qi_eager_threshold")
 }
 
+/// Set `qi.profile` value to the given one.
```
(Also for `smt_qi_profile_freq`:) please actually explain the effect of this setting. Reading the description here a user might not realise there is any connection to Z3, since you don't mention it and they might not be aware that Z3 has such a setting.

Aurel300 commented on prusti-utils/src/config.rs:
```diff
@@ -1017,3 +1043,29 @@ pub fn cargo_command() -> String {
+    read_setting("show_ide_info")
+}
+
+/// Very similar to no_verify but should only be used by prusti-assistant
```
"Very similar"—so is there a difference? What is it? Would it make more sense to make "verification mode" into an enum, which would contain IDE modes as well as the "no verify" mode and maybe other test modes?

cedihegi replied:
I will add a comment and describe it in the dev-guide, but yes there are some important differences. 
It is needed because of issue #1261. My solution for that problem was to generate a fake error when we skip verification.
The problem with `no_verify` is that it's value is not just what we pass via the command line. If no-verify is set to false by the user, it will still evaluate to true when compiling dependencies, and if I start throwing fake errors at that point, this will not work. 
If issue #1261 is resolved, this flag and the fake errors will no longer be necessary.
Let me know if you still think creating this enum you described would be a good idea, but I don't think that this skip_verification variable should be in there since it's probably temporary.

Aurel300 commented on prusti-smt-solver/Cargo.toml:
```diff
@@ -12,7 +12,8 @@ doctest = false # and no doc tests
 
 [dependencies]
 futures = "0.3"
+log = { version = "0.4", features = ["release_max_level_info"] }
```
I assume this change is not needed: there is no actual change to `prusti-smt-solver` in this PR.

Aurel300 commented on prusti-server/src/verification_request.rs:
```diff
@@ -56,9 +56,11 @@ impl ViperBackendConfig {
                     // to #unspecified
                     format!(
-                        "smt.qi.eager_threshold={} model.partial={}",
+                        "smt.qi.eager_threshold={} model.partial={} smt.qi.profile={} smt.qi.profile_freq={}",
```
(See comment in `config.rs`.)

Aurel300 commented on prusti-server/tests/basic_requests.rs:
```diff
@@ -61,10 +60,21 @@ where
         ),
     };
 
+    let get_result_type = |acc: VerificationResultKind, msg: ServerMessage| async move {
```
Why is this a closure, why `async`, why `move`?

Aurel300 commented on prusti-server/tests/basic_requests.rs:
```diff
@@ -61,10 +60,21 @@ where
+            PrustiClient::verify(SERVER_ADDRESS.clone(), request)
+                .await
+                .fold(VerificationResultKind::Failure(vec![]), get_result_type)
+                .await
```
This is not very pretty… Is there an easy way to `await` all the stream items to get an iterator (or a collection), then just use [`Iterator::find`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) to find the termination message and [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) to get a failure if there was none?

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -3387,6 +3388,14 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             ).with_span(call_site_span)?
         };
+        // store spans of items of the contracts for prusti-assistant
+        self.store_contract_spans(
```
Can this only be called when some config flag is enabled?

Aurel300 commented on prusti-viper/src/encoder/errors/position_manager.rs:
```diff
@@ -39,7 +39,6 @@ impl<'tcx> PositionManager<'tcx>
         let span = span.into();
         let pos_id = self.next_pos_id;
         self.next_pos_id += 1;
-        trace!("Register position id {} for span {:?} in {:?}, ", pos_id, span, def_id);
```
Why delete?

Aurel300 commented on prusti-viper/src/ide/vsc_span.rs:
```diff
@@ -0,0 +1,41 @@
+                expansion: None,
+            })
+        } else {
+            None
```
How common is this? Do we reasonably expect every span that is passed here to be a valid span? If so, this could maybe be an `Err` rather than `None`, or even a panic? If it is an error (`Result` return type), the method should be named `try_from_span`.

cedihegi replied:
I think panicking here should be reasonable.

Aurel300 commented on prusti-viper/src/ide/encoding_info.rs:
```diff
@@ -0,0 +1,45 @@
+        call_span: Span,
+        contracts_spans: Vec<Span>,
+        source_map: &SourceMap
+    ) -> Option<Self> {
```
A function called `new` should not return an `Option`. Please check the comment in `VscSpan` for whether this even needs to be an `Option`/`Result` at all.

Aurel300 commented on prusti-viper/src/ide/encoding_info.rs:
```diff
@@ -0,0 +1,45 @@
+use super::vsc_span::VscSpan;
+
+/// stores the spans of a calls contracts.
+/// obtained during encoding
```
Please reword and expand.

Aurel300 commented on prusti-viper/src/ide/ide_verification_result.rs:
```diff
@@ -0,0 +1,23 @@
+use viper::VerificationResult;
+
+// since the existing verification result
+// is not as trivially passed in json
```
Reword, expand.

Aurel300 commented on prusti-viper/src/ide/ide_verification_result.rs:
```diff
@@ -0,0 +1,23 @@
+}
+
+impl IdeVerificationResult {
+    pub fn from_res(res: &VerificationResult) -> Self {
```
Should this be a [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) implementation?

cedihegi replied:
yes!

Aurel300 commented on prusti-viper/src/ide/vsc_span.rs:
```diff
@@ -0,0 +1,41 @@
+    file_name: String,
+    is_primary: bool,
+    label: Option<()>,
+    expansion: Option<()>,
```
Can you explain these two fields? Why are they not just `bool`s?

cedihegi replied:
The idea was to allow easy serialization here, and deserialization in typescript, where spans of this form are already parsed. (this is the form they have when emitting compiler messages with spans).

`label` and `expansion` are both optional fields in the type defined in prusti-assistant, so putting in a boolean would not work. Option fields with value None, when serialized, will result in no field. But since I don't use them at all I can just delete them.

Aurel300 commented on prusti-server/src/server_message.rs:
```diff
@@ -0,0 +1,22 @@
+use viper::VerificationResult;
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub enum ServerMessage {
```
Can you add a comment explaining what this is, where it comes from? Each variant should also have a brief explanation.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -3528,6 +3537,73 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                }
+            })
+            .collect();
+        let mut result = vec![]; 
```
It looks to me like you have a bunch of `DefId`s, for each of them you end up calling `get_def_span`. How about: create an iterator chain of all of the `DefId`s, then `map` `get_def_span` over it once, then `collect` into a `Vec` rather than append to it like in the following lines…

cedihegi replied:
I somewhat adjusted it, but it's still not very pretty.
I tried to generate a chain of iterators, but I did not manage to make this prettier at all. 
For the pledges I still have to merge them after finding the spans, so I can not include them. For the others it's also not quite as simple as it might look since there are 2 vecs and 2 `Option<Option<DefId>>`. I put them all into a vector now before querying the spans. I hope that's alright.

Aurel300 commented on prusti-viper/src/verifier.rs:
```diff
@@ -74,204 +86,318 @@ impl<'v, 'tcx> Verifier<'v, 'tcx> {
+        // if we are in an ide, we already emit the errors asynchronously, otherwise we wait for
+        // all of them in order to sort them
+        if !config::show_ide_info() {
+            prusti_errors.sort();
```
The comment about sortedness (just before `let mut prusti_errors`) should be here.

Aurel300 commented on prusti-viper/src/verifier.rs:
```diff
@@ -74,204 +86,318 @@ impl<'v, 'tcx> Verifier<'v, 'tcx> {
-        (program_name, request)
-    });
-    if let Some(server_address) = config::server_address() {
+    /// Returns a list of (program_name, verification_requests) tuples.
```
Returns verification requests?

Aurel300 commented on prusti-viper/src/verifier.rs:
```diff
@@ -74,204 +86,318 @@ impl<'v, 'tcx> Verifier<'v, 'tcx> {
+                        )
+                        .emit(&self.env.diagnostic);
+                    }
+                    match result.result_type {
```
The `match` arms here are quite deeply nested. Does it make sense to extract handling for the cases into separate methods? It would also be good to be able to quickly see how much of the scope is actually needed for each (i.e. just the result kind + `prusti_errors`? `self`? `quantifier_instantiations`?).

Aurel300 commented on prusti-server/src/client.rs:
```diff
@@ -4,58 +4,76 @@
+        server_address: S,
+        request: VerificationRequest,
+    ) -> impl Stream<Item = ServerMessage> {
+        // TODO: do proper error handling
```
How "improper" is it? For example, is it just resilience against the server unexpectedly quitting? If we panic in such a case that might be ok.

jthomme1 replied:
Atm, we panic if the server address is invalid and if the connection is torn down/fails somehow unexpectedly. However, this was basically the same before: just that there this method returned a result which was `.expect`ed.

Aurel300 commented on prusti-server/src/process_verification.rs:
```diff
@@ -120,11 +266,129 @@ pub fn process_verification_request<'v, 't: 'v>(
+                        let stripped = no_pref
+                            .strip_suffix("-aux")
+                            .or(no_pref.strip_suffix("_precondition"))
+                            .unwrap_or(no_pref);
```
Where do these alternatives come from?

jthomme1 replied:
I did not investigate too much, but as far as I remember they are introduced by Z3

Aurel300 commented on prusti-server/src/process_verification.rs:
```diff
@@ -120,11 +266,129 @@ pub fn process_verification_request<'v, 't: 'v>(
+                    let pos_string =
+                        jni.get_string(jni.unwrap_result(obj_wrapper.call_toString(pos)));
+                    let pos_id_index = pos_string.rfind('.').unwrap();
+                    let pos_id = pos_string[pos_id_index + 1..].parse::<u64>().unwrap();
```
Why is the position handling different here and in the quantifier instantiations message? Can they share code?

jthomme1 replied:
The handling is different due to the different message types. In the QuantifierInstantiationsMessage, we get a Z3 quantifier name (e.g. a string `prog.l234-aux`) while in the QuantifierChosenTriggersMessage, we get an actual `Positioned` jObject, where we extract the position id from the string of the position (because we needed less Java boiler plate code, but maybe it would be better to get it from the actual line member).

Aurel300 commented on prusti/src/ide_helper/call_finder.rs:
```diff
@@ -0,0 +1,124 @@
+                                .push((defpath_resolved, resolved_def_id, sp));
+                        } else {
+                            // in this case we want both
+                            self.called_functions
+                                .push((defpath_resolved, resolved_def_id, sp));
+                            self.called_functions
+                                .push((defpath_unresolved, method_def_id, sp));
```
Should the entries in `called_functions` also contain some indication of what kind of method it is? For example, the IDE could display "extern spec for trait method X" vs "extern spec for trait impl method X"?

Aurel300 commented on prusti/src/ide_helper/call_finder.rs:
```diff
@@ -0,0 +1,124 @@
+            }
+            ExprKind::Binary(..) | ExprKind::AssignOp(..) | ExprKind::Unary(..) => {
+                let resolve_res = self.resolve_expression(expr);
+                // this will already fail for standard addition
```
As in, for example, `i32` addition, because it is a built-in operation?

cedihegi replied:
yes exactly. I will reword it to built-in operations instead of standard addition.

Aurel300 commented on prusti/src/ide_helper/compiler_info.rs:
```diff
@@ -0,0 +1,128 @@
+    queried_source: Option<String>,
+    // additionally this will contain:
+    // function_calls:
+    // ... we'll see
```
?

cedihegi replied:
left-over comment

Aurel300 commented on prusti/src/ide_helper/compiler_info.rs:
```diff
@@ -0,0 +1,128 @@
+        let queried_source = query_signature::collect_queried_signature(env.tcx(), &fncalls);
+        Self {
+            procedure_defs: procs,
+            function_calls: fncalls,
```
Rename `procs` to `procedure_defs` and `fncalls` to `function_calls`.

Aurel300 commented on prusti/src/ide_helper/compiler_info.rs:
```diff
@@ -0,0 +1,128 @@
+                defid,
+                span: VscSpan::from_span(&sp, source_map).unwrap(),
+            })
+            .collect();
```
```suggestion
            .collect::<Vec<_>>();
```
(bit nicer than explicitly annotating `fncalls` with a type hint)

Aurel300 commented on prusti/src/ide_helper/compiler_info.rs:
```diff
@@ -0,0 +1,128 @@
+    pub span: VscSpan,
+}
+
+impl Serialize for ProcDef {
```
You can add [`#[serde(skip)]`](https://serde.rs/field-attrs.html#skip) to the `defid` field to skip it, instead of manually implementing the trait.

Aurel300 commented on prusti/src/ide_helper/query_signature.rs:
```diff
@@ -0,0 +1,432 @@
+};
+
+/// data structure to represent the code we want to generate
+/// since we didnt manage to this with syn etc..
```
Reword, expand. The reference to `syn` only makes sense to us because we discussed it…

Aurel300 commented on prusti/src/ide_helper/query_signature.rs:
```diff
@@ -0,0 +1,432 @@
+            }
+            DefKind::AssocFn => {
+                // this will be None for traits
+                let impl_defid_opt = tcx.impl_of_method(defid);
```
Why the local?

Aurel300 commented on prusti/src/ide_helper/query_signature.rs:
```diff
@@ -0,0 +1,432 @@
+                    }
+                    None => {
+                        // function is a Trait (or something else?)
+                        // are there other cases for AssocFns?
```
I don't think so.

Aurel300 commented on prusti/src/ide_helper/query_signature.rs:
```diff
@@ -0,0 +1,432 @@
+                // not sure if this is actually completely correct.
+                // can a single generic type have the same traitbound
+                // more than once with different arguments?
+                // I would imagine yes.. may be a todo
```
Like `T: Foo<i32> + Foo<u32>`? Yes, this is possible.

Commit:
Implement some comments (untested)

Commit:
Fix Z3 arguments

Commit:
Fix test and formatting error

Commit:
Refactor / Fix according to code reviews

Commit:
Format and move old comment to right location

Commit:
Address reviews and add comments

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Improve displaying of generated extern spec block for trait

Commit:
Fix some quantifier reporting weirdness

Also add some debug statements.

Commit:
Make QI etc. also work for existential quantifiers

Commit:
Check if current package is primary package before throwing fake_error

In 2 places we throw fake errors to avoid caching of successes when we
should not. This lead to problems with local dependencies, since they are
verified too and the fake_error ended up being thrown too early. This
bug is resolved now.

Commit:
Separate trait bounds from generic args and handle bad case

First of all we separated trait bounds from generic args, since for
example for a function signature within an impl block, the trait bounds
at the function level can still refer to the generic args defined in the
impl block.
Secondly, we had a problem if a generic had duplicate traitbounds, with
different generics, e.g.: `T: Foo<u32> + Foo<String>`, this is now
handled correctly, under the assumption that the predicates occurr in a
certain order, i.e. projections come after their respective trait. So
far this seems to be always the case, if it's not I don't see how this
problem can be solved.
